File: fsharp/Simplify.FSharp/Simplify.fs
namespace Simplify

type Point =
    | Point2d of float * float
    | Point3d of float * float * float

module Simplify =
    let private mapPoint point =
        match point with
        | Point2d (x, y) -> (x, y, 0.0)
        | Point3d (x, y, z) -> (x, y, z)

    let internal squareDistance left right =
        let squareDiff (xl, yl, zl) (xr, yr, zr) =
            ((xl - xr) ** 2.0)
            + ((yl - yr) ** 2.0)
            + ((zl - zr) ** 2.0)

        squareDiff (mapPoint left) (mapPoint right)

    let internal segmentSquareDistance point segmentStart segmentEnd =
        let sq a = a * a

        let (x, y, z) = mapPoint point
        let (startX, startY, startZ) = mapPoint segmentStart
        let (endX, endY, endZ) = mapPoint segmentEnd

        let deltaX = endX - startX
        let deltaY = endY - startY
        let deltaZ = endZ - startZ

        let result rx ry rz =
            let dx = x - rx
            let dy = y - ry
            let dz = z - rz
            sq dx + sq dy + sq dz

        if deltaX <> 0.0 || deltaY <> 0.0 || deltaZ <> 0.0 then
            let t =
                ((x - startX) * deltaX
                 + (y - startY) * deltaY
                 + (z - startZ) * deltaZ)
                / (sq deltaX + sq deltaY + sq deltaZ)

            if t > 1.0
            then result endX endY endZ
            else result (startX + deltaX * t) (startY + deltaY * t) (startZ + deltaZ * t)
        else
            result startX startY startZ

    let rec internal simplifyRadialDistance points tolerance =
        match points with
        | []
        | [ _ ] -> points
        | point :: rest ->
            point
            :: simplifyRadialDistance
                (rest
                 |> List.skipWhile (fun p -> squareDistance p point < tolerance))
                   tolerance

    let rec internal simplifyDouglasPeucker points tolerance =
        match points with
        | []
        | [ _ ]
        | [ _; _ ] -> points
        | _ ->
            let first = points.Head
            let last = points |> List.last

            let (index, distance) =
                points
                |> List.mapi (fun index point ->
                    if point = first || point = last
                    then (index, -1.0)
                    else (index, segmentSquareDistance point first last))
                |> List.maxBy snd

            if distance > tolerance then
                ((simplifyDouglasPeucker (points |> List.take (index + 1)) tolerance)
                 |> List.rev
                 |> List.skip 1
                 |> List.rev)
                @ (simplifyDouglasPeucker (points |> List.skip (index)) tolerance)
            else
                [ first; last ]

    let simplify points tolerance highQuality =
        match points with
        | []
        | [ _ ]
        | [ _; _ ] -> points
        | _ ->
            let squaredTolerance = tolerance * tolerance

            if highQuality
            then simplifyDouglasPeucker points squaredTolerance
            else simplifyDouglasPeucker (simplifyRadialDistance points squaredTolerance) squaredTolerance

File: scala/src/main/scala/Simplify.scala
import scala.math.pow

abstract class Point

case class Point2d(x: Double, y: Double) extends Point

case class Point3d(x: Double, y: Double, z: Double) extends Point

object Simplify {
  private type Coordinates = (Double, Double, Double)

  private implicit def fromPoint(p: Point): Coordinates = p match {
    case p: Point2d => fromPoint2d(p)
    case p: Point3d => fromPoint3d(p)
  }

  private implicit def fromPoint2d(p: Point2d): Coordinates = (p.x, p.y, 0)

  private implicit def fromPoint3d(p: Point3d): Coordinates = (p.x, p.y, p.z)

  private def removeLast[A](list: List[A]) = list
    .reverse
    .drop(1)
    .reverse

  private def squareDistance(left: Point, right: Point) = {
    def squareDiff(l: Coordinates, r: Coordinates) = {
      val (xl, yl, zl) = l
      val (xr, yr, zr) = r
      pow(xl - xr, 2) + pow(yl - yr, 2) + pow(zl - zr, 2)
    }

    squareDiff(left, right)
  }

  private def segmentSquareDistance(point: Point, segmentStart: Point, segmentEnd: Point) = {
    def sq(a: Double) = a * a

    val (x, y, z): Coordinates = point
    val (startX, startY, startZ): Coordinates = segmentStart
    val (endX, endY, endZ): Coordinates = segmentEnd

    val (deltaX, deltaY, deltaZ) = (endX - startX, endY - startY, endZ - startZ)

    def result(rx: Double, ry: Double, rz: Double) = {
      val dx = x - rx
      val dy = y - ry
      val dz = z - rz
      sq(dx) + sq(dy) + sq(dz)
    }

    if (deltaX != 0 || deltaY != 0 || deltaZ != 0) {
      val t = ((x - startX) * deltaX
        + (y - startY) * deltaY
        + (z - startZ) * deltaZ) / (sq(deltaX) + sq(deltaY) + sq(deltaZ))
      if (t > 1)
        result(endX, endY, endZ)
      else
        result(startX + deltaX * t, startY + deltaY * t, startZ + deltaZ * t)
    } else
      result(startX, startY, startZ)
  }

  private def simplifyRadialDistance(points: List[Point], tolerance: Double): List[Point] = points match {
    case List() | List(_) => points
    case point :: rest => point :: simplifyRadialDistance(rest.dropWhile(p => squareDistance(p, point) < tolerance), tolerance)
  }

  private def simplifyDouglasPeucker(points: List[Point], tolerance: Double): List[Point] = points match {
    case List() | List(_) | List(_, _) => points
    case _ =>
      val first = points.head
      val last = points.last

      val (index, distance) = points
        .zipWithIndex
        .map { case (point, index) => if (point == first || point == last) (index, -1.0) else (index, segmentSquareDistance(point, first, last)) }
        .maxBy(_._2)

      if (distance > tolerance)
        removeLast(simplifyDouglasPeucker(points.take(index + 1), tolerance)) ++ simplifyDouglasPeucker(points.drop(index), tolerance)
      else
        List(first, last)
  }

  def apply(points: List[Point], tolerance: Double, highQuality: Boolean): List[Point] = points match {
    case List() | List(_) | List(_, _) => points
    case _ =>
      val squaredTolerance = tolerance * tolerance
      if (highQuality)
        simplifyDouglasPeucker(points, squaredTolerance)
      else
        simplifyDouglasPeucker(simplifyRadialDistance(points, squaredTolerance), squaredTolerance)
  }
}
